### PR TITLE
JAX: Add section mlp-implementation & numerical stability

### PR DESCRIPTION
**Pull request by AnirudhDagar**
Add JAX implementation for sections 5.2 and 5.4.

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_multilayer-perceptrons/mlp-implementation.md [hunk 1]
```diff
@@ -198,6 +230,22 @@ class MLP(d2l.Classifier):
...
+```{.python .input}
+%%tab jax
+class MLP(d2l.Classifier):
+    num_outputs: int
+    num_hiddens: int
+    lr: float
+
+    @nn.compact
+    def __call__(self, X):
+        X = X.reshape((X.shape[0], -1)) # flatten
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/mlp-implementation.md:242]: Add two space right before inline comments.
Capitalize the first letter

File: chapter_multilayer-perceptrons/mlp-implementation.md [hunk 2]
```diff
@@ -198,6 +230,22 @@ class MLP(d2l.Classifier):
...
 ```
 
+```{.python .input}
+%%tab jax
+class MLP(d2l.Classifier):
+    num_outputs: int
+    num_hiddens: int
+    lr: float
+
+    @nn.compact
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/mlp-implementation.md:240]: Have we used this decorator before?

File: chapter_multilayer-perceptrons/numerical-stability-and-init.md [hunk 3]
```diff
@@ -192,6 +207,17 @@ for i in range(100):
 print('after multiplying 100 matrices\n', M.numpy())
 ```
 
+```{.python .input}
+%%tab jax
+get_new_key = lambda: jax.random.PRNGKey(d2l.get_seed())  # generate PRNG keys
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/numerical-stability-and-init.md:212]: Capitalize the first letter of comments

**Review by astonzhang — APPROVED** [review]

File: chapter_multilayer-perceptrons/mlp-implementation.md (hunk 2 above)
Comment (AnirudhDagar) [reply chapter_multilayer-perceptrons/mlp-implementation.md:240]: No, this is a more compact way of defining a network in flax, we should explain this in the text. But right now, I'm just adding the code. Do you want me to modify the text as well?

Comment (astonzhang) [reply chapter_multilayer-perceptrons/mlp-implementation.md:240]: For better readability, please add text to explain it

Comment (AnirudhDagar) [reply chapter_multilayer-perceptrons/mlp-implementation.md:240]: Done

File: chapter_linear-classification/softmax-regression-concise.md [hunk 4]
```diff
@@ -70,6 +70,14 @@ We use a `Flatten` layer to convert the 4th order tensor `X`
 by keeping the dimension along the first axis unchanged.
 :end_tab:
 
+:begin_tab:`jax`
+Flax allows users to write the network class in a more compact way
+using `@nn.compact` dectorator. With `@nn.compact`, one
+can simply write all network logic inside a single “forward pass”
+method, without needing to define the standard `setup` method in
```
Comment (astonzhang) [inline chapter_linear-classification/softmax-regression-concise.md:77]: What is the `setup` method？

Comment (AnirudhDagar) [reply chapter_linear-classification/softmax-regression-concise.md:77]: https://github.com/d2l-ai/d2l-en/blob/36dccad4c5c0793974028316fc7ded2ee5151920/d2l/jax.py#L192

Comment (astonzhang) [reply chapter_linear-classification/softmax-regression-concise.md:77]: It's far away and it's hard to refer to code comments (not prominent) in previous sections. Can we add more descriptions of the setup method here?
